File: drivers/typescript/tests/query.test.ts
import { describe, it, expect } from 'vitest';
import { R, QueryBuilder, TableQuery, DbQuery } from '../src/query';

describe('QueryBuilder', () => {
  const r = new R();

  it('should create a table query', () => {
    const query = r.table('users');
    expect(query).toBeInstanceOf(TableQuery);
    const term = query.toTerm();
    expect(term).toEqual([15, [14, 'test'], 'users']);
  });

  it('should create a db query', () => {
    const query = r.db('mydb');
    expect(query).toBeInstanceOf(DbQuery);
    const term = query.toTerm();
    expect(term).toEqual([14, 'mydb']);
  });

  it('should chain filter', () => {
    const query = r.table('users').filter({ active: true });
    const term = query.toTerm();
    expect(term[0]).toBe(39); // FILTER
  });

  it('should chain limit', () => {
    const query = r.table('users').limit(10);
    const term = query.toTerm();
    expect(term[0]).toBe(42); // LIMIT
    expect(term[2]).toBe(10);
  });

  it('should chain skip', () => {
    const query = r.table('users').skip(5);
    const term = query.toTerm();
    expect(term[0]).toBe(43); // SKIP
    expect(term[2]).toBe(5);
  });

  it('should chain orderBy', () => {
    const query = r.table('users').orderBy('name');
    const term = query.toTerm();
    expect(term[0]).toBe(41); // ORDER_BY
  });

  it('should chain count', () => {
    const query = r.table('users').count();
    const term = query.toTerm();
    expect(term[0]).toBe(86); // COUNT
  });

  it('should chain get', () => {
    const query = r.table('users').get('abc123');
    const term = query.toTerm();
    expect(term[0]).toBe(16); // GET
    expect(term[2]).toBe('abc123');
  });

  it('should chain insert', () => {
    const query = r.table('users').insert({ name: 'John' });
    const term = query.toTerm();
    expect(term[0]).toBe(17); // INSERT
  });

  it('should chain update', () => {
    const query = r.table('users').get('abc').update({ name: 'Jane' });
    const term = query.toTerm();
    expect(term[0]).toBe(18); // UPDATE
  });

  it('should chain delete', () => {
    const query = r.table('users').get('abc').delete();
    const term = query.toTerm();
    expect(term[0]).toBe(19); // DELETE
  });

  it('should create dbList query', () => {
    const query = r.dbList();
    const term = query.toTerm();
    expect(term).toEqual([59]);
  });

  it('should create dbCreate query', () => {
    const query = r.dbCreate('mydb');
    const term = query.toTerm();
    expect(term).toEqual([57, 'mydb']);
  });

  it('should create tableCreate query', () => {
    const query = r.db('test').tableCreate('users');
    const term = query.toTerm();
    expect(term[0]).toBe(60); // TABLE_CREATE
  });

  it('should chain changes', () => {
    const query = r.table('users').changes();
    const term = query.toTerm();
    expect(term[0]).toBe(152); // CHANGES
  });

  it('should chain multiple operations', () => {
    const query = r.table('users')
      .filter({ active: true })
      .orderBy('name')
      .skip(10)
      .limit(5);
    const term = query.toTerm();
    // Should be nested: LIMIT(SKIP(ORDER_BY(FILTER(TABLE, ...), ...), ...), ...)
    expect(term[0]).toBe(42); // LIMIT (outermost)
  });
});

File: drivers/go/query_test.go
package gothinkdb

import (
	"testing"
)

func TestNewR(t *testing.T) {
	r := NewR()
	if r == nil {
		t.Fatal("NewR returned nil")
	}
}

func TestTableQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users")
	if q == nil {
		t.Fatal("Table returned nil")
	}
	term := q.ToTerm()
	if len(term) != 2 {
		t.Fatalf("expected 2 elements in term, got %d", len(term))
	}
	if term[0] != TermTable {
		t.Errorf("expected term[0] = %d (TABLE), got %v", TermTable, term[0])
	}
}

func TestDbQuery(t *testing.T) {
	r := NewR()
	q := r.DB("mydb")
	if q == nil {
		t.Fatal("DB returned nil")
	}
	term := q.term
	if len(term) != 2 {
		t.Fatalf("expected 2 elements in term, got %d", len(term))
	}
	if term[0] != TermDB {
		t.Errorf("expected term[0] = %d (DB), got %v", TermDB, term[0])
	}
	if term[1] != "mydb" {
		t.Errorf("expected term[1] = 'mydb', got %v", term[1])
	}
}

func TestFilterChain(t *testing.T) {
	r := NewR()
	q := r.Table("users").Filter(map[string]interface{}{"active": true})
	term := q.ToTerm()
	if term[0] != TermFilter {
		t.Errorf("expected FILTER term, got %v", term[0])
	}
}

func TestLimitChain(t *testing.T) {
	r := NewR()
	q := r.Table("users").Limit(10)
	term := q.ToTerm()
	if term[0] != TermLimit {
		t.Errorf("expected LIMIT term, got %v", term[0])
	}
	if term[2] != 10 {
		t.Errorf("expected limit 10, got %v", term[2])
	}
}

func TestSkipChain(t *testing.T) {
	r := NewR()
	q := r.Table("users").Skip(5)
	term := q.ToTerm()
	if term[0] != TermSkip {
		t.Errorf("expected SKIP term, got %v", term[0])
	}
	if term[2] != 5 {
		t.Errorf("expected skip 5, got %v", term[2])
	}
}

func TestGetQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").Get("abc123")
	term := q.ToTerm()
	if term[0] != TermGet {
		t.Errorf("expected GET term, got %v", term[0])
	}
	if term[2] != "abc123" {
		t.Errorf("expected id 'abc123', got %v", term[2])
	}
}

func TestInsertQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").Insert(map[string]interface{}{"name": "John"})
	term := q.ToTerm()
	if term[0] != TermInsert {
		t.Errorf("expected INSERT term, got %v", term[0])
	}
}

func TestUpdateQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").Get("abc").Update(map[string]interface{}{"name": "Jane"})
	term := q.ToTerm()
	if term[0] != TermUpdate {
		t.Errorf("expected UPDATE term, got %v", term[0])
	}
}

func TestDeleteQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").Get("abc").Delete()
	term := q.ToTerm()
	if term[0] != TermDelete {
		t.Errorf("expected DELETE term, got %v", term[0])
	}
}

func TestCountQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").Count()
	term := q.ToTerm()
	if term[0] != TermCount {
		t.Errorf("expected COUNT term, got %v", term[0])
	}
}

func TestDBListQuery(t *testing.T) {
	r := NewR()
	q := r.DBList()
	term := q.ToTerm()
	if len(term) != 1 || term[0] != TermDBList {
		t.Errorf("expected [59], got %v", term)
	}
}

func TestDBCreateQuery(t *testing.T) {
	r := NewR()
	q := r.DBCreate("mydb")
	term := q.ToTerm()
	if len(term) != 2 || term[0] != TermDBCreate || term[1] != "mydb" {
		t.Errorf("expected [57, 'mydb'], got %v", term)
	}
}

func TestTableCreateQuery(t *testing.T) {
	r := NewR()
	q := r.DB("test").TableCreate("users")
	term := q.ToTerm()
	if term[0] != TermTableCreate {
		t.Errorf("expected TABLE_CREATE term, got %v", term[0])
	}
}

func TestChangesQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").Changes()
	term := q.ToTerm()
	if term[0] != TermChanges {
		t.Errorf("expected CHANGES term, got %v", term[0])
	}
}

func TestChainedQuery(t *testing.T) {
	r := NewR()
	q := r.Table("users").
		Filter(map[string]interface{}{"active": true}).
		OrderBy("name").
		Skip(10).
		Limit(5)
	term := q.ToTerm()
	// Should be nested: LIMIT(SKIP(ORDER_BY(FILTER(TABLE, ...), ...), ...), ...)
	if term[0] != TermLimit {
		t.Errorf("expected outermost LIMIT, got %v", term[0])
	}
}

func TestUseDB(t *testing.T) {
	r := NewR()
	r.Use("production")
	q := r.Table("users")
	term := q.ToTerm()
	// Table now just uses [TABLE, "users"] without DB reference
	if term[0] != TermTable {
		t.Errorf("expected TABLE term, got %v", term[0])
	}
	if term[1] != "users" {
		t.Errorf("expected table 'users', got %v", term[1])
	}
}

func TestIndexOperations(t *testing.T) {
	r := NewR()

	// IndexCreate
	q := r.Table("users").IndexCreate("email")
	term := q.ToTerm()
	if term[0] != TermIndexCreate {
		t.Errorf("expected INDEX_CREATE, got %v", term[0])
	}

	// IndexList
	q = r.Table("users").IndexList()
	term = q.ToTerm()
	if term[0] != TermIndexList {
		t.Errorf("expected INDEX_LIST, got %v", term[0])
	}

	// IndexDrop
	q = r.Table("users").IndexDrop("email")
	term = q.ToTerm()
	if term[0] != TermIndexDrop {
		t.Errorf("expected INDEX_DROP, got %v", term[0])
	}
}
